action-server/tests/test_job_send_reminder.py

```diff
@@ -0,0 +1,163 @@
+import asyncio
+from unittest import TestCase
+from unittest.mock import patch
+
+from aiohttp import web
+from aiohttp.test_utils import TestServer, unused_port
+
+from db.reminder import Reminder
+from jobs.send_reminders import run
+
+EN = "en"
+FR = "fr"
+
+HASHIDS_SALT = "abcd1234"
+HASHIDS_MIN_LENGTH = 4
+
+ENV = {
+    "REMINDER_ID_HASHIDS_SALT": HASHIDS_SALT,
+    "REMINDER_ID_HASHIDS_MIN_LENGTH": str(HASHIDS_MIN_LENGTH),
+}
+
+REMINDER_1 = Reminder(
+    id=1,
+    timezone="America/Toronto",
+    attributes={"first_name": "Bob", "phone_number": "15145551234", "language": FR},
+)
+
+REMINDER_2 = Reminder(
+    id=2,
+    timezone="America/Toronto",
+    attributes={"first_name": "Lilly", "phone_number": "15145554567", "language": EN},
+)
+
+
+class TestJobSendReminder(TestCase):
+    def _setUp(
+        self,
+        mock_session_factory,
+        reminders=[REMINDER_1, REMINDER_2],
+        fail_request=False,
+        start_server=True,
+    ):
+        loop = asyncio.get_event_loop()
+        self.server = dict()
+        port = dict()
+
+        self.server[EN] = FakeCoreServer(fail_request)
+        self.server[FR] = FakeCoreServer(fail_request)
+
+        if start_server:
+            port[EN] = loop.run_until_complete(self.server[EN].start())
+            port[FR] = loop.run_until_complete(self.server[FR].start())
+        else:
+            port[EN] = unused_port()
+            port[FR] = unused_port()
+
+        self.core_endpoints = {
+            EN: f"127.0.0.1:{port[EN]}",
+            FR: f"127.0.0.1:{port[FR]}",
+        }
+
+        mock_session_factory.return_value.query.return_value.filter.return_value.all.return_value = (
+            reminders
+        )
+
+    def tearDown(self):
+        try:
+            loop = asyncio.get_event_loop()
+            loop.run_until_complete(self.server[EN].stop())
+            loop.run_until_complete(self.server[FR].stop())
+        except AttributeError:
+            pass
+
+    def test_missing_environment_variables(self):
+        with self.assertRaises(expected_exception=Exception):
+            run()
+
+    @patch.dict("os.environ", ENV)
+    @patch("jobs.send_reminders.session_factory")
+    def test_with_environment_variables(self, mock_session_factory):
+        self._setUp(mock_session_factory)
+
+        sent, errored = run(core_endpoints=self.core_endpoints)
+
+        self.assertCountEqual(sent, [REMINDER_1.id, REMINDER_2.id])
+        self.assertCountEqual(errored, [])
+
+    @patch("jobs.send_reminders.session_factory")
+    def test_send_reminders(self, mock_session_factory):
```

Comment:
There's no test to check if the timing conditions to send a reminder are respected. For all we know, the job could send all reminders each time it is executed. I haven't calculated the amount of efforts to put in to test it but it's kind of an important feature...

Reply:
Yep I know, that's the part I have no idea how to test. We can't unit test it so it has to be with an integration test that will add fake data to the db and the effort required to do this quite high.
If you have other ideas let me know...

Reply:
I thought you might be able to unit test with mocks but looking at the code a bit more I realize where the difficulty lies, and no, I can't think of a solution. :/ We'll leave it like this and hope no one breaks it for now I guess.